Time integrators

sanguinariojoe opened the pull request:
**THIS PR DEPENDS ON https://github.com/FloatingArrayDesign/MoorDyn/pull/224**

Well, I am afraid this will require quite a bit of discussion on several of the changes I implemented:

* Fixed the Adams-Bashforth schemes [here](https://github.com/FloatingArrayDesign/MoorDyn/commit/5a03001c08f736d9a246dd5a7ed22d318d5f6a6e). I suppose this does not require any kind of discussion at all.
* Added a CFL ([Courant–Friedrichs–Lewy](https://en.wikipedia.org/wiki/Courant%E2%80%93Friedrichs%E2%80%93Lewy_condition)) option, in a such a way the user can either define the time step (dtM) or the CFL factor (cfl). He can even define both, so the most restrictive is considered.
* Removed the Z presevation check on the catenary solver (see [here](https://github.com/core-marine-dev/MoorDyn/commit/6570d0281036b060219cc45546a98398f94d610c)). To be honest, I do not remember if we added that, or it was always there. Anyway, if we want that back, we need to use the relative error instead. Otherwise, the errors for big depths will fall inside that condition (600 meters seems enough)
* Seabed is only considered now on the catenary solver when one of the line-ends is close enough to the bottom. This fix the initialization in many situations, and I do not think it is worsening the results on any context.
* Added Local-timestep versions of the Euler and the Adams-Bashforth time schemes. Tests on the lEuler have been created
* I substituted the previous initial condition by a stationary solver, which is essentially the same than before but with an infinitely large dissipation (indeed the trick is that velocity is completely removed). The main benefit is that the new solver is unconditionally stable (with the booster it might get unstable for a while though). Some other benefits are that it is able to converge to a way more precise solution. On the other hand, the null velocity might means that the solution would take significantly longer to be reached, specially when the catenary solution fails. Anyway, since the user still have the possibility of saving and loading, I do not see much benefit on the previous approach, which can still be used anyway.

sanguinariojoe commented:
Hey guys! I uploaded a Newmark scheme. An Average Constant Acceleration (ACA) one to be more specific, with relaxation, which is able to remain stable with significantly larger time steps.

@mattEhall and @RyanDavies19 , it would be great if you can review this and give feedback.

RyanDavies19 commented:
@sanguinariojoe just keeping you in the loop, Matt and I have both been pretty busy with other work so we haven't gotten the chance to dig too much into this yet. It's still on our radar, and hopefully I'll be able to get to it soon. Currently I am working on adding vortex induced vibrations to MoorDyn.

sanguinariojoe commented:
> Currently I am working on adding vortex induced vibrations to MoorDyn

Cool! I was told to avoid this on the past because apparently those are engineered out, so our digital twins shall not take them into account

sanguinariojoe commented:
Hey @RyanDavies19 !

This branch is evolving quite much... I am indeed never using dev branch anymore.

- The stationary solution is way more powerful than the previous Drag increased Initial Condition generator. Probably there is room for even further improvements.
- The implicit schemes (specially the midpoint) are clearly outperforming the explicit ones.
- I added auto-mem-checks, which are really handy.
- I fixed several errors

I really think we should start addressing this before both branches start diverging too much

RyanDavies19 commented:
@sanguinariojoe agreed. I will be reviewing today now that I have some time available for MDC work. Thank you for including the docs, that is very helpful. My biggest concerns are making sure that 1) we retain backwards compatibility in the form of input files, and 2) the new input file is well documented (which it seems you've already done a lot of). Many of the questions I see in the OpenFAST issues/forum are from confusions between the two different code versions so keeping things clear will be important.

sanguinariojoe commented:
Well, actually i want to document it a bit further. The discussion on the
relaxation factor is unfinished

On Thu, 13 Jun 2024, 18:31 RyanDavies19, ***@***.***> wrote:

> @sanguinariojoe <https://github.com/sanguinariojoe> agreed. I will be
> reviewing today now that I have some time available for MDC work. Thank you
> for including the docs, that is very helpful. My biggest concerns are
> making sure that 1) we retain backwards compatibility in the form of input
> files, and 2) the new input file is well documented (which it seems you've
> already done a lot of). Many of the questions I see in the OpenFAST
> issues/forum are from confusions between the two different code versions so
> keeping things clear will be important.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/FloatingArrayDesign/MoorDyn/pull/182#issuecomment-2166172012>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAMXKKGTIY4VDLNX2Z7AO23ZHHCPRAVCNFSM6AAAAABB2HBDPWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCNRWGE3TEMBRGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

RyanDavies19 commented:
@sanguinariojoe okay that would be good. I'll add comments in the code where some slight changes might be needed

RyanDavies19 commented on docs/inputs.rst:
```diff
@@ -165,7 +165,8 @@ parenthesis). As such, they are all optional settings, although some of them (su
 
- - dtM (0.001) – desired mooring model time step (s)
+ - dtM (3.402823e+38) – desired mooring model maximum time step (s)
+ - CFL (0.5) – desired mooring model maximum CFL factor
```
These CFL notes are helpful, but need to be in the V2 section (this is V1). Additionally they should be under the line "The following MoorDyn-C options are not supported by MoorDyn-F:"

sanguinariojoe replied:
Fixed on https://github.com/FloatingArrayDesign/MoorDyn/pull/182/commits/5c3a404c3b2ee7809ebda94754e611c4c4f30691

RyanDavies19 commented on docs/tschemes.rst:
```diff
@@ -0,0 +1,314 @@
+============
+.. _tschemes:
+
+MoorDyn 2 is deployed with several time schemes, with different features,
```
"MoorDyn-C version 2 is deployed..."

sanguinariojoe replied:
Fixed on https://github.com/FloatingArrayDesign/MoorDyn/pull/182/commits/1bb95b347bb7c25d949dfde5ce6b5e448a2a073a

sanguinariojoe commented on docs/tschemes.rst:
```diff
@@ -0,0 +1,314 @@
+
+MoorDyn 2 is deployed with several time schemes, with different features,
+strengths and weaknesses.
+
```
Fixed on https://github.com/FloatingArrayDesign/MoorDyn/pull/182/commits/e4b95e01ad0a8e090fb2e896e91054764f82f40c

sanguinariojoe commented:
I rebased this PR on https://github.com/FloatingArrayDesign/MoorDyn/pull/224, so that PR shall be merged first

sanguinariojoe commented:
Nice! I think #224 can be merged. Anyway the verifications against openfast
dev have to be made manually

On Mon, 24 Jun 2024, 18:27 RyanDavies19, ***@***.***> wrote:

> @sanguinariojoe <https://github.com/sanguinariojoe> there should be a PR
> up to OpenFAST/dev tomorrow afternoon with updated reg tests for comparison
> using #224 <https://github.com/FloatingArrayDesign/MoorDyn/pull/224>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/FloatingArrayDesign/MoorDyn/pull/182#issuecomment-2186963554>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAMXKKDK4LUHYEOQB4AIWO3ZJBCHXAVCNFSM6AAAAABB2HBDPWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCOBWHE3DGNJVGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

RyanDavies19 commented:
@sanguinariojoe I agree it looks good to go. OpenFAST dev is a few weeks away from merging into main so we should be able to get the new reg tests in by then. If not it's a simple fix to point to the dev branch.

RyanDavies19 commented:
@sanguinariojoe Added a few more docs reviews. I will open a PR to your branch with the fixes. Overall though they look really good and are very helpful. Thanks for putting them together!

RyanDavies19 commented on docs/inputs.rst:
```diff
@@ -548,16 +552,20 @@ indicate which version has that as the default. As such, they are all optional s
  - g (9.81): The gravity acceleration (m/s^2)
  - rho (1025): The water density (kg/m^3)
- - WtrDpth (0.0): The water depth (m). In MoorDyn-F the bathymetry file path can be inputted here.
+ - WtrDpth (0.0): The water depth (m). In MoorDyn-C the bathymetry file path can be inputted here.
```
Resolved

RyanDavies19 commented on docs/inputs.rst:
```diff
@@ -179,6 +179,11 @@ default value provided in parentheses) is:
  - ThreshIC (0.001) – convergence threshold for IC generation, acceptable relative difference 
    between three successive fairlead tension measurements (-)
 
+The internal time step is first taken from the dtM option, and then adjusted
```
Resolved

RyanDavies19 commented on source/MoorDyn2.cpp:
```diff
@@ -87,7 +86,8 @@ moordyn::MoorDyn::MoorDyn(const char* infilename, int log_level)
-  , dtM0(0.001)
+  , dtM0((std::numeric_limits<real>::max)())
+  , cfl(0.5)
   , dtOut(0.0)
```
Resolved

RyanDavies19 commented:
See https://github.com/OpenFAST/r-test/issues/127 for updated reg tests

sanguinariojoe commented on source/MoorDyn2.cpp:
```diff
@@ -302,8 +297,30 @@ moordyn::MoorDyn::Init(const double* x, const double* xd, bool skip_ic)
+	// Compute the timestep
+	for (auto obj : LineList)
+		dtM0 = (std::min)(dtM0, obj->cfl2dt(cfl));
+	for (auto obj : PointList)
```
It is indeed. But in the future that might change so they might have their own requirements for stability or accuracy. A formal analysis shall be made

sanguinariojoe commented:
> See [OpenFAST/r-test#127](https://github.com/OpenFAST/r-test/pull/127) for updated reg tests

Hey @RyanDavies19 , the new tests do not work to me... In OpenFAST I mean. It seems that all the drivers named like `BodiesAndRods_driver.dvr` shall be replaced by `md_driver.inp`

RyanDavies19 commented:
@sanguinariojoe Agreed lets merge. I talked with Matt and that's fine. Apologies for confusion with https://github.com/OpenFAST/r-test/pull/127, I should have opened that as a draft. It is a work in progress.

sanguinariojoe commented:
Hurray! Thanks guys!
